### PR TITLE
V1 refactor

### PR DESCRIPTION
**Pull request by tonykaralis**

### CODE REVIEW

**Comment by PoLaKoSz** [comment]
Hi!

I had some free time so i cooked something: [feature/simplified-requests](https://github.com/PoLaKoSz/TMdbEasy/tree/feature/simplified-requests).

This is how i would append the API key to the requested URL without exposing it everywhere.
I know there is a lot of changes that is because i am on Ubuntu so i couldn't build .NET Framework projects but i push a new commit soon to make it easier to see my actual changes.

**Comment by tonykaralis** [comment]
I agree that there is a simpler way of hiding the api key no doubt, your pull request addresses this and that was my original plan. However in the context of a web app, the usage may differ. A web app may have all movie requests scoped to a client in which case we would want to give the developer the ability to use a client specific api key.

Hey,

The idea was to bake it in with the refactor so nobody is forced to use this feature, but its available if required. Adding it in after the fact would just mean more refactoring more breaking changes. Plus, I was planning on building something at a later date that could use this library with different api keys, not sure if that will ever happen 😄 

The api factory is syntactic sugar, it can be injected into a controller or a service which requires access to several apis without needing a massive constructor, in fairness that can be removed.

**Comment by tonykaralis** [comment]
Whats most important to me here is developer experience and clean code. If we removed the ability to provide a user specific api key it would make the code a lot simpler but also takes away a powerful feature.

Aside from KISS is there any other reason why we wouldn't keep the per user apikey feature in?

**Comment by tonykaralis** [comment]
In fact, your changes have prompted an idea. Let me make a few changes, i'll share a commit with you.

**Comment by PoLaKoSz** [comment]
Yes this make sense. I don't have too much experience with C# web apps so there is a big chance that i didn't think some use cases here. If i am not mistaken C# web apps could share resources (variables) between multiple requests not like PHP. The current implementation could help remove the overhead to initialize some classes (`TmdbEasyClientv3`, `HttpClient`, is there anything else?), but in my opinion this won't be a bottleneck for a web app.
I read somewhere that developers should focus on writing readable code and after that fine tune the performance. If You decide to set the API key only in the `TmdbEasyOptions` for now but some point in the future support to set API key for each request You could still inherit from a class (like now) without breaking change (i think).

**Comment by tonykaralis** [comment]
Setting the api key on the `TmdbEasyOptions` will be the default. Its easy to configure and easy for devs to understand and it follows a well known pattern.

The current implementation will indeed cut down on the initialisation of `TmdbEasyClientv3` and also supports DI out of the box. HttpClient is a tricky one and needs to be treated with care. It has been built and optimised to be reused for the entire life of an application and in fact can lead to serious performance degradation if scoped to a request.
> HttpClient is intended to be instantiated once and re-used throughout the life of an application. Instantiating an HttpClient class for every request will exhaust the number of sockets available under heavy loads. This will result in SocketException errors. Below is an example using HttpClient correctly. see [ms docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netcore-3.1). for more info on this.
I ran a test with just 20 requests and it left about 20 sockets taken long after the requests had ended.

A better way is to use the IHttpClientFactory but this is only available in .Net Core and this is a .NET standard library.

As you mention sharing resources like the apikey across requests is no problem at all. 

I 100% agree with this.
> I read somewhere that developers should focus on writing readable code and after that fine tune the performance.

Your changes have prompted an idea and I think i can make the code even simpler and still allow a user api key per request without exposing it to each individual api method. Let me give this a think and Ill share a commit with you over the next day.

Thanks for the feedback, great to have you back.

**Comment by PoLaKoSz** [comment]
My other idea woud be to remove the `string language = "en"` from the methods and just add a `public string Language { get; set ; }` to the `TmdbEasyOptions` and append the content for each request (i don't think TMDb will reject a request just because an endpoint doesn't have an optional language parameter.

**Comment by PoLaKoSz** [comment]
HttpClient could be injected to the `TmdbEasyClientv3` ctor so it will make it easier for developers to use only one even if there is more TmdbEasyClientv3 creater for some reason.
Glad if i could help :smile:

**Comment by tonykaralis** [comment]
Regarding the language idea, this would imply that every request would be scoped to a single language. I don't think this is ideal, since an application with different users in different countries might want to query for different languages.

The TmdbEasyOptions are meant to be created once and set at application startup. These options wont change for the lifetime of the app. This is the typical MS configuration pattern which works well in almost all apps I have designed and built at work.

**Comment by PoLaKoSz** [comment]
Yes, You are right about the language for web apps. But if i use it for only one language setting it only in for example in the `TmdbEasyOptions` and forget about it would be easier than passing it for each request.

**Comment by tonykaralis** [comment]
This PR closes #18

**Comment by tonykaralis** [comment]
@PoLaKoSz this commit [c9e2d95](https://github.com/tonykaralis/TMdbEasy/pull/17/commits/c9e2d95cdb4d31ba414bb6f5964c4dbd3f3e38db) has all the necessary changes to address #18  and #19 

In summary:
1. The request building has been completely separated from the client and api code
2. Users have the choice to easily set an apikey globally via TmdbEasyOptions
3. Users have the option to override the global api key by providing an api key to each individual api request
4. Users have the choice to easily set an default language globally  via TmdbEasyOptions
5. Users have the option to override the default language by providing a language to each individual api request
6. ApiFactory has been removed as unnecessary syntactic sugar
7. Request building has been made slightly more robust to avoid basic errors, this might need tweaking in the future if via usage we find any bugs
8. Removes all custom request classes to expose the exact properties that are required vs optional for each api method

Thanks for you feedback on this, hopefully this makes the developer experience a bit smoother

**Comment by PoLaKoSz** [comment]
- I like repos with this folder structure (i follow it myself):

```
|- src
|    |- TmdbEasy.csproj
|
|- tests
|    |- Unit
|    |   |- Unit.csproj
|    |
|    |- Integration
|        |- Integration.csproj
|
|- README.md
```
What do You think?
Namespaces: `TmdbEasy.Tests.Unit` and `TmdbEasy.Tests.Integration`?

- I really like this Request builder approach, very neat! Personally i even make it simpler by removing `TmdbEasyClientv3.GetBaseUrl()`, `TmdbEasyClientv3.GetApiKey()` and `TmdbEasyClientv3.GetDefaultLanguage()` methods and pass the `TmdbEasyOptions` into the `RequestHandler` and `Request` ctor to decrease the parameters count (and remove the 3 mentioned `TmdbEasyClientv3` methods).
- Just to make the code even more readable (fluentish) i would rename the `RequestHandler.ExecuteRequestAsync()` to `ExecuteAsync`.
- You sad that the `TmdbEasyOptions` should be set only once. The class does not indicate this (public setters). I know its bit more code but like i suggested before, set the properties in the ctor instead (i would also change the visibility of the properties to internal too).
- I would simplify our life to this:
```c#
namespace TmdbEasy.Configurations
{
    public class TmdbEasyOptions
    {
        private const string TmdbUrl3 = "http://api.themoviedb.org/3";
        private const string TmdbUrl3Ssl = "https://api.themoviedb.org/3";

        public TmdbEasyOptions(bool useSsl, string apiKey, string defaultLanguage)
        {
            this.BaseUri = useSsl ? TmdbUrl3Ssl : TmdbUrl3;
            this.ApiKey = apiKey;
            this.DefaultLanguage = defaultLanguage;

        }

        internal string BaseUri { get; }

        internal string ApiKey { get; }

        internal string DefaultLanguage { get; }
    }
}
```
and now the `GetBaseUrl()` from `TmdbEasyClientv3` could be removed too. I think specifying the BaseAddress for the HttpClient was unnecessary because the query contains the full path, am i right?
- I see this everywhere but i just can't understand this Async suffix. I see when this naming convention could be usefull, [here is an example](https://stackoverflow.com/a/47810905/7306734), but i never used it because when i build upon a HttpClient every call will be async.

**Comment by tonykaralis** [comment]
I edited your post with some numbering so it would make my comments easier to follow:

1. Yes I use this at work as well, tried something different here but I don't like it either
2. Agree this can be an easy way of simplifying the code
3. Yes that sounds good 👍 
4. Typically you would hold on to that class after its been set at app startup so not really an issue, but I'm happy to have that change made.
5. Yes you are right, this can be removed if we make the change in 2.
6. There are varying opinions on this but I tend to follow the pattern of.
>If a public method is Task-returning and is asynchronous in nature (as opposed to a method that is known to always execute synchronously to completion but still returns a Task for some reason), it should have an “Async” suffix.  

I think Id prefer to leave 6. as is.

How are you doing for time? I still need to finish off a couple apis before merging v1. How about i start on 1. and 4. and leave 2. 3. 5. for you as they are closely related. If possible you could submit PR onto the v1 branch.

Let me know if you have time or else i can just knock those out. Thanks for the help

**Comment by PoLaKoSz** [comment]
I code 2, 3 and 5 right now, and do the first after that because i think that will cause a merge conflict.

**Comment by tonykaralis** [comment]
I missed your comment, I have made the changes to address 1.

**Comment by PoLaKoSz** [comment]
Can You revert or hard reset the commit? Even GitHub says in #20 :smile: :

> These conflicts are too complex to resolve in the web editor

**Comment by tonykaralis** [comment]
I will revert after I have reviewed your PR. I will then apply my commits back on top of your PR once its merged

**Comment by PoLaKoSz** [comment]
Thanks!

**Comment by tonykaralis** [comment]
@PoLaKoSz thanks for help on this.

Releasing v1.0.0-alpha1 as we haven't got a full test suite for Movie and television api. Once that's done will bump to v1.0.0

@VPKSoft Will work on testing the television api within the week, I may have introduced a couple bugs.

**Comment by PoLaKoSz** [comment]
Glad that i was able to help :)

**Comment by PoLaKoSz** [comment]
Nothing stops the user to send a request without setting an API key (forgot to set in the Options and in the parameter too). Fix in #24

<hr>

To send one request the setup is painfully long:
``` c#
var deserializer = new NewtonSoftDeserializer();
var client = new TmdbEasyClientv3(deserializer, new TmdbEasyOptions());
var requestHandler = new RequestHandler(client);
var endPoint = new ConfigApi(requestHandler);
var jobs = await endPoint.GetJobsAsync();
```

Solution:
- i would swap the order of the `TmdbEasyClientv3` ctor parameters (i just want to supply an API key via Options and don't care how the library works internally)
- then i would make the `IJsonDeserializer` an optional parameter
- i would define clearly the responsibility of the `ITmdbEasyClient` interface: Provide an entry point for the library and nothing more:

``` c#
public interface ITmdbEasyClient
{
    IChangesApi Changes { get; }
    ICollectionApi Collection { get; }
    ICompaniesApi Companies { get; }
    IConfigApi Config { get; }
    ICreditApi Credit { get; }
    IMovieApi Movie { get; }
    INetworksApi Networks { get; }
    IReviewApi Review { get; }
    ITelevisionApi Television { get; }
}

public class TmdbEasyClientv3 : ITmdbEasyClient
{
    private static readonly  string _tmdbUrl3 = "http://api.themoviedb.org/3/";
    private static readonly string _tmdbUrl3Ssl = "https://api.themoviedb.org/3/";

    public TmdbEasyClientv3(TmdbEasyOptions options)
        : this(options, new NewtonSoftDeserializer())
    {
    }

    public TmdbEasyClientv3(TmdbEasyOptions options, IJsonDeserializer jsonDeserializer)
    {
        if (options == null)
            throw new ArgumentNullException(nameof(options));

        var httpClient = new HttpClient()
        {
            BaseAddress = new Uri(options.UseSsl ? _tmdbUrl3Ssl : _tmdbUrl3)
        };

        var requestHandler = new RequestHandler(options, httpClient, jsonDeserializer);

        Changes = new ChangesApi(requestHandler);
        Collection = new CollectionApi(requestHandler);
        Companies = new CompaniesApi(requestHandler);
        Config = new ConfigApi(requestHandler);
        Credit = new CreditApi(requestHandler);
        Movie = new MovieApi(requestHandler);
        Networks = new NetworksApi(requestHandler);
        Review = new ReviewApi(requestHandler);
        Television = new TelevisionApi(requestHandler);
    }

    public IChangesApi Changes { get; }
    public ICollectionApi Collection { get; }
    public ICompaniesApi Companies { get; }
    public IConfigApi Config { get; }
    public ICreditApi Credit { get; }
    public IMovieApi Movie { get; }
    public INetworksApi Networks { get; }
    public IReviewApi Review { get; }
    public ITelevisionApi Television { get; }
}
```
- i would change the visibility to `internal` for every class in the the `TmdbEasy.Apis` namespace (one less public namespace)
- now we can make `RequestHandler` internal so we won't need the `IRequestHandler` interface anymore. I made a slight change in the RequestHandler ctor params: now it accepts 3 classes. This class could be responsible to send a request (this is why i moved the HttpClient here) to an URL and parse its content accordingly:

``` c#
internal class RequestHandler
{
    private readonly HttpClient _httpClient;
    private readonly IJsonDeserializer _jsonDeserializer;
    private readonly TmdbEasyOptions _options;

    internal RequestHandler(
        TmdbEasyOptions options,
        HttpClient httpClient,
        IJsonDeserializer jsonDeserializer)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _options = options;
        _jsonDeserializer = jsonDeserializer;
    }

    internal Request CreateRequest()
    {
        return new Request(_options);
    }

    internal async Task<TmdbEasyModel> ExecuteAsync<TmdbEasyModel>(Request request)
    {
        string url = request.GetUri();
        if (string.IsNullOrEmpty(url) || string.IsNullOrWhiteSpace(url))
            throw new ArgumentException("Generated url is invalid (null or empty)!", nameof(request));

        string json = await _httpClient.GetStringAsync(url)
            .ConfigureAwait(false);

        return _jsonDeserializer.DeserializeTo<TmdbEasyModel>(json);
    }
}
```

Result:
``` c#
var tmdb = new TmdbEasyClientv3(new TmdbEasyOptions("apiKey"));
var jobs = await tmdb.Config.GetJobsAsync();
```

<hr>

Yes, i removed the `ApiVersion GetVersion();` method from the `ITmdbEasyClient` interface because we only support v3 for now so once again KISS principle :smile:.

_`RequestHandler` ctor could be simplified probably with this:_
``` c#
public class TmdbEasyClientv3 : ITmdbEasyClientv3, ITmdbEasyClient {}

internal interface ITmdbEasyClient
{
    TmdbEasyOptions Options { get; }
    HttpClient HttpClient { get; }
    IJsonDeserializer JsonDeserializer { get; }
}
```
_... and just pass this interface intead of 3 parameters (and probably the future `TmdbEasyClientv4` could inherit from this interface too)._

**Comment by tonykaralis** [comment]
Hi Tom,

- Thanks for the #24 just some failing tests to look at.
- I am also OK with removing the version, not that it complicates anything but no problem there.

Regarding the rest, this refactor is far too large and in my opinion is not a SOLID design. I can see where you are coming from in that you want to have a simple API, however you are looking at this only from a desktop apps POV. The code changes made above would not play nicely with a web application and definitely not with Dependency Injection. [see ms docs for a starter](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1#:~:text=ASP.NET%20Core%20supports%20the,controllers%20in%20ASP.NET%20Core.) The client now depends on all the APIS as well. 

- There is clear violation of the SRP principle (The client is now doing too many things)
    - exposing APIS
    - new-ing classes up (taken them on as a dependency)
    - adds complexity when testing
    - cannot swap out the API implementations without changing the client
    - cannot swap out the API implementations at runtime either
    - creating HttpClient and RequestHandler per request as this would simple not be able to be shared across httpcontexts
- Violates Open-closed principle -> each time we add a new API we have to change the client
- 2 constructors generally bad design (What are we telling the user with this)?

The above changes make it super easy for a desktop to spin up a client and make a call, but it makes things a whole lot more complicated and very messy for WebApps.

The API as it is, is very clear IMHO. It does not violate SRP, complies with Open-Closed and supports DI. It also decouples the implementation from the contract and makes it very easy for the dev to set things up as well as use, allows for API implementations to be replaced at runtime and without changing the client

If you look at [.NET Core Config](https://github.com/tonykaralis/TMdbEasy#net-core-configuration) you'll see the setup is a one liner. If you look at [Usage](https://github.com/tonykaralis/TMdbEasy#usage) you'll see the usage is super simple as well. The DI container takes care of everything.

In the context of a desktop app, you may not have all the DI buts and this is where, instead of refactoring the whole library to suit a single use case, you would:

1. create a wrapper which would encapsulate your setup code into a single reusable method:

> var deserializer = new NewtonSoftDeserializer();
var client = new TmdbEasyClientv3(deserializer, new TmdbEasyOptions());
var requestHandler = new RequestHandler(client);
var endPoint = new ConfigApi(requestHandler);
var jobs = await endPoint.GetJobsAsync();

2. create a factory that would return the types you need. 
>  public IChangesApi Changes { get; }
    public ICollectionApi Collection { get; }
    public ICompaniesApi Companies { get; }
    public IConfigApi Config { get; }
    public ICreditApi Credit { get; }
    public IMovieApi Movie { get; }
    public INetworksApi Networks { get; }
    public IReviewApi Review { get; }
    public ITelevisionApi Television { get; }

In my opinion the above should happen within the client application(not within TmdbEasy) as it is extremely specific to that app. 

is there a specific app that is using this library? Perhaps i can help you with looking at that. Typically desktops can make use of DI as well, but with the solution I mention above, it wouldn't have to.

**Comment by PoLaKoSz** [comment]
Thank You for the detailed answer i really appreciate it. Yes, You are right, i have much less experience with web apps thats for sure but i have a few questions.
My understanding is that API key and the language could be set for each request, so this means TmdbEasyClientv3 can be registered as a singleton in the DI container. When the DI creates the TmdbEasyClientv3 it will initialize (currently 9) `...Api` classes and it will live while the web server is running. _When i typed my last sentece i realized that these 9 endpoints created inside the TmdbEasyClientv3 class so DI can't access them_. ... But is is true? 

**_Few hours later ..._ :smile:**

[After i rewrite completly](https://github.com/PoLaKoSz/TMdbEasy/blob/feature/shorter-setup-code/src/TmdbEasyServiceCollectionExtensions.cs) (just a proof of concept) the `TmdbEasyServiceCollectionExtensionsAddTmdbEasy()`  method i was able to [successfully run an ASP .NET Core WebAPI server](https://github.com/PoLaKoSz/TMdbEasy/blob/feature/shorter-setup-code/samples/web-api/Controllers/TmdbController.cs) and get result back from TMDb.

<hr>

> - There is clear violation of the SRP principle (The client is now doing too many things)

You mean `TmdbEasyClientv3`? Its only responsible to initialize the library to be usable in the future.

>   - exposing APIS

Yes, guides the developer how can access all endpoints without looking up the documentation.

>   - new-ing classes up (taken them on as a dependency)

Yes, right now only 2 , and 9 required classes. But `HttpClient` should be injected through the ctor not as is now.

> - adds complexity when testing

What do You mean? You can test everything except the relative generated URLs but they covered with the Integrations tests (just not explicitly). In fact test setup even shorter.

_I left the `Request` class untouched because URL generation is really important, but i would probably make that class internal too._

> - cannot swap out the API implementations without changing the client

I don't understand this point, sorry.

> - cannot swap out the API implementations at runtime either

I don't understand this point either, sorry. Can You give me an example?

> - creating HttpClient and RequestHandler per request as this would simple not be able to be shared across httpcontexts

HttpClient created inside the `TmdbEasyClientv3` class only once in the ctor. It will be created once (even with ASP .NET Core because its registered as singleton).

> - Violates Open-closed principle -> each time we add a new API we have to change the client

Yes i agree with You. The same thing could happen when TMDb decide to add a new endpoint for example `movie/{id}/actors`. The person who will implement it probably add a new method to the `IMovieApi` interface so the principle will be break or is there any other option? Extension method?

> - 2 constructors generally bad design (What are we telling the user with this)?

I never heard something like this. For me it means there are multiple **good** ways to initialize a class that is all.

**Comment by tonykaralis** [comment]
No problem at all, its fun having discussions and looking at new approaches, learning new things etc.

Before I respond to questions, I'd like to mention singletons. Some people love them others hate them, I use them only if there is a compelling reason to.

I registered the TmdbEasyClient as singleton as it uses an HttpClient which needs to exist for the life of the application,so it make sense that this is a singleton. All the apis however, do not need to exist the entire life of the application.

In your sample, you have registered the TmdbEasyClient as a singleton and then all of its dependencies as well (9 apis)). So you have to ask yourself _"do the api classes really need to live the life of the application?"_ 

The answer is yes, because **the design forces** them too, they are unable to exist any other way. And this is an immediate flag that something might be wrong with the design. What if somebody only wants a movie-api, they have to live with the fact that there are 8 other apis new-ed up and alive for no reason at all.

1. There is clear violation of the SRP principle (The client is now doing too many things)
> You mean TmdbEasyClientv3? Its only responsible to initialise the library to be usable in the future.
Looking at it again, perhaps it is only doing one thing.  If you have multiple constructors, then it may look like your object is doing different things, depending on how its constructed. And in this case it is. Probably a better way would be to have the JsonDeserializer an optional parameter. I still dont agree with what the class does or how it does it, but will comment on the rest of your questions.

2. exposing APIS
> Yes, guides the developer how can access all endpoints without looking up the documentation.
This can be done via a factory, and id argue it _should_ be done by a factory. I will see if I can get something working for you. On of the reason i am not hiding all these classes is to allow for them to be replaced by the caller, should the caller not like the way something is handled. This makes the library super flexible. 

For example: My app might need to do something every time the `RequestHandlers` `.GetUri()` method is called (lets just say for the sake of simplicity), all Id need to do, is implement the `IRequesthandler` interface. i register it with DI and off I go, The library works without needing any changes.

PS. At the very least the dev should read the README file. There we have listed the 9 apis.

3. new-ing classes up (taken them on as a dependency)
> Yes, right now only 2 , and 9 required classes. But HttpClient should be injected through the ctor not as is now.
This will only get more and more.  will discuss open-closed later on.

4. adds complexity when testing
> What do You mean? You can test everything except the relative generated URLs but they covered with the Integrations tests (just not explicitly). In fact test setup even shorter.
Ignore this for now, Testing was referring to just the testing of the TmdbEasyClient class but after looking at it it might be ok.

5. cannot swap out the API implementations without changing the client
> I don't understand this point, sorry.
I mntioned this above in point 2 as well. The user of the library might want to change the implementation of an API. With the current version they can do this very easily, since nothing in the library depends on the implementation(classes) of the APIS. So if I wanted to write my own NetworkAPI or RequestHandler or TmdbEasyClient and still have it work with all the rest of the library, I can. But with your changes, this would not be possible. The TmdbEasyClient depends directly on all the implementations.

6. cannot swap out the API implementations at runtime either
> I don't understand this point either, sorry. Can You give me an example?
This follows on from 5. At runtime using a flag and DI we can change which version of an implementation we are using based on the user, the environment or anything else that we might need. You might say _"but the library is super simple that will never happen"_ but that is besides the point, the design is decoupled and everything is programmed to an interfac, decoupled and interchangeable.

7. creating HttpClient and RequestHandler per request as this would simple not be able to be shared across httpcontexts
> HttpClient created inside the TmdbEasyClientv3 class only once in the ctor. It will be created once (even with ASP .NET Core because its registered as singleton).
yeap, after you wrote a new serviceCollection extension method then I understood how it would be used.
8. Violates Open-closed principle -> each time we add a new API we have to change the client
> You absolutly right. I don't think i eveer was able to follow the SOLID principles 100%, but i allways try as closly as possible. On the other hand we have to realize this library is a wrapper so we depend on an external dependency which could change any time. In my opinion declaring a clean entry point and addig new endpoints if necessary will go against the open-closed rule, but it would be a minor change and for me writing code as a developer is more fun then reading docs on how to use a library (even worse, when i have to read the source).
The goal is to use SOLID as a guide, so for sure we dont have to apply it 100%. Sometimes we just cant, but here we have a choice and we can follow the OpenClosed principle. The cass should be open for extension (add a new method) but closed for modification (not changing an existing method everytime we add a new feature(api in this instance)).

I still believe the best way to give you your siple developer experience is by leaving everything public as is so devs can change things if they wish. Then we can add a factory which can aid in providing the entry point you wish.

2 constructors generally bad design (What are we telling the user with this)?
> I never heard something like this. For me it means there are multiple good ways to initialize a class that is all.
If you have multiple constructors, then it may look like your object is doing different things, depending on how its constructed. Generally, your object should only be doing 1 thing.

Maybe you can have a crack at implementing a factory on top of the master branch?? I will too as we can see what comes of it?

**Comment by PoLaKoSz** [comment]
Thank You for taking the time i hope i am not full defensive here, i try to reason with as much objectiveness as possible :smile:

Given this code:
``` c#
class Car {}
class Engine
{
      public static Engine Instance() { /* .. */ }
}
```
I never used singletons as [described here](https://csharpindepth.com/Articles/Singleton) and i think registering `Engine` class with DI as a singleton and calling a static method for an `Engine` instance is two different thing.
With singleton pattern developers hide the `Car` class dependencies because usually in the ctor will call `_engine = Engine.Instance();` and this is a really bad practise.
On the other hand even if i register a class singleton with DI and pass the `Engine` dependency through the `Car`'s ctor i won't hide the dependency so this is a good practise. If i register as singleton is just allocates less memory and when resolving the class second time DI, it will return the object faster.

You stated lot of time that You want to support web apps out of the box. Lets assume a website has an endpoint which uses this library to get a movie details and this endpoint called 10 times per second (this web app doesn't use anything to cache the responses so it will call TMDb every time.
- If we register every `I...Api` as a singleton the DI will won't sweat because after the second request it will return the same `MovieApi` instance
- If we register every `I...Api` as transient than the DI will resolve 20 new `MovieApi` instance in two seconds

You can see if request/seconds gets higher it gets ever worse for no good reason. No `...Api` holds any user specifc state , thanks the option to set the API key and language on a per-request basis so no need to create each time we need a `...Api`. And yes, You are right, in the memory there will be 8 unused objects, but compared to the **10 new object/second** isn't it the best solution?

> For example: My app might need to do something every time the RequestHandlers .GetUri() method is called (lets just say for the sake of simplicity), all Id need to do, is implement the IRequesthandler interface. i register it with DI and off I go, The library works without needing any changes.

All right, i can imagine that there is a use case to get every requested URL. You could expose `IRequestInterface` then as You mentioned the user **have to** create a new class, inject it, etc **or** You could expose an event on the `TmdbEasyClientv3` (for example `BeforeRequest` where this event will return the generated URL from `RequestHandlers .GetUri()`. No new class, no injecting, no hussle just simple event subscription for the user. Only we have to implement it once internally without any breaking change and the need for any other library user who needs the same. What do You think about this?

> Maybe you can have a crack at implementing a factory on top of the master branch?? I will too as we can see what comes of it?

Sorry, but i don't understand it, can You rephrase it for me, please? Thanks!

**Comment by tonykaralis** [comment]
No problem at all, I dont take your questions and interest as defensive at all ;)

With regards to singletons, I am not referring to the singleton design pattern, as you mentioned that can be bad practice. What I am referring to is registering dependencies with DI as singletons only if you need to. 

You should only really be registering singletons in DI for 2 reasons:
1.  if the class directly depends on and uses resources which can be limited e.g. sockets, database connections etc 
2.  if state matters.

> If i register as singleton is just allocates less memory and when resolving the class second time DI, it will return the object faster.

This should not be taken into consideration and should not be affecting your design. In most cases the performance hit cannot be noticed and is not an issue. We have several production web apps with hundreds of users and hundreds of services and there is no performance hit. A small ASP.NET Core kestrel server on a 1gb vm handles 1000's per second without even breaking a sweat.

Having 9 ...API instances alive even if we only need one is bad design but not because of memory resource. It is the **concept of creating services when they are not needed** that is wasteful. 
Imagine you are sick and you go to the doctor. The doctor checks you out and says you need treatment. So he prescribes a fancy medicine that costs 1000£ a tablet and says you only need 1 tablet. So you go to the pharmacy with the prescription and the pharmacist gives you a box of 9 pills and says "unfortunately even though you only need 1 you have to buy 9, the other 8 will just have to sit around be there even if you don't need them". To some the cost wont matter, to others it will. In this example we are the pharmacist and its our job to give 1 pill when 1 is requested. We don't care about the cost, we care about doing correctness. 

In the current version, the only class that needs resource that is limited(sockets) is the `TmdbEasyClient` since its the only class that really _needs_ the HttpClient. From what I have gathered, I think you understand this, but let me know if not.

All the remaining classes depend on the `ITmdbEasyClient` interface and know nothing about the client. All of these classes do not need to be registered as singletons. In fact, it is best practice that they are not registered as singletons. Neither of them need access to a limited resource and neither of them need to maintain some sort of state. 

Transcient and scoped service registration is generally better when the resource is used temporarily. In a typical web app, the user will make a request to the server, the server will spin up an ..API instance, do the work, once the request returns the server then disposes the instance(DI manages the lifetime of the instances).

Also you didn't comment on the fact that users may want to replace the ...API instance with their own versions or opt between versions. Generally your decision hides everything away and leave 0 room for extension or modification from the user.

Regarding the factory, I will get back to you on this with another comment, I need to pop out now.

**Comment by PoLaKoSz** [comment]
Okey, i am gonna accept what You sad about registering singleton dependencies.

I forgot to write answer to this:

> PS. At the very least the dev should read the README file. There we have listed the 9 apis.

Yes, they should be, but there could be a problem when the docs out of sync with the code. Intellisense hopefully never lies, so remembering only the `TmdbEasy.TmdbEasyClientv3` class can help explore the whole library. Yesterday when i made the sample web app i encountered the same problem:

**README.md**
``` c#
public void ConfigureServices(IServiceCollection services)
{
    // Adds TmdbEasy to your app
    services.AddTmdbEasy();
}
```
https://github.com/tonykaralis/TMdbEasy/blob/1b09de580f6dcb241904f4c9f6b551520b617890/src/TmdbEasyServiceCollectionExtensions.cs#L9-L12

The `AddTmdbEasy` has a required `TmdbEasyOptions` parameter.

**Comment by tonykaralis** [comment]
Docs is another discussion. The README is just a simple bare bones intro to the Web App. I left the options out on purpose but you make a good point, I should probably include the options so the example is complete.

Trying to over simplify a library to save someone from going in to the docs is probably not a good thing to do. Think about most libraries, you need to at some point read something to figure out how things work, if you need advanced options etc.

Typically for any library most devs would go to the docs before using the library (at least this is what I advise for professional programming). In our case the library is small so we don't really have a  lot to put in the docs.

I will look at improving the README slightly but I will focus on keeping it short, simple and minimal. In theory, if I make a few tweaks to the Services section and add the options in, that should be enough for someone to get up and running.

Do you think a 1 pager wiki could be worth adding? I think its overkill.. whats your thoughts on that?

**Comment by PoLaKoSz** [comment]
Interesting. I like using and writing libraries that are ready to use without any setup if possible. My opinion is when i want to access an API, i don't want to know how the library works i just want to call the correct method end get a correct response that i could use in my own app. Creating 2 additional (`RequestHandler`, `...Api`) classes to do that, injecting into each other and then using the library is not what is in my mind (as a mainly ConsoleApp developer :smile:) ... 2-3 line top. If i am not satisfied with the library then i will read more about it. But this is my opinion, thank You for letting me know yours.

I think the README close to perfect and i don't think any Wiki page needed (i would disable the menu in the Settings too), only a small code sample for ConsoleApp devs missing in my opinion.

**.NET Core Configuration** section
``` diff
- that dont use an IOC container.
+ that don't use an IOC container.
```

_Just to keep the code sample short_
``` diff
public void ConfigureServices(IServiceCollection services)
{
-    services.Configure<CookiePolicyOptions>(options =>
-    {
-        options.CheckConsentNeeded = context => true;
-        options.MinimumSameSitePolicy = SameSiteMode.None;
-    });

    // Adds TmdbEasy to your app
     services.AddTmdbEasy();

-    // Other service registrations
-    services.AddAuthentication();

-    services.AddMvc();

-    services.AddAntiforgery();
}
```

The `I...Api` interface list under the **Supported Services** could be hrefs to the actual interfaces to easily navigate to each of them and take a look to all available members.

`IUserService` never used
``` diff
public class MovieController : Controller
{
     private readonly IMovieApi _movieApi;
-    private readonly IUserService _userService;
    
-    public HomeController(IMovieApi movieApi, IUserService userService)
+    public HomeController(IMovieApi movieApi)
     // ...
}

**Comment by tonykaralis** [comment]
Right Tom, I have been giving it some thought and understand your need to hide some of the implementation classes. I agree that we can trim this down but not by much. 

I have the following suggestion:
1. The APIS must remain public and all the implementations as well. These need to be known, they need to be managed by the user, so new-ing them up should not be the responsibility of this library. We should not be responsible for the lifetime of these objects either. We can only advise what the suggested scope should be, offer a default and the rest is up to the user. If someone wants DI fine, and if a console dev doesn't want DI then also fine. We only give them what they absolutely need. That leaves us with 2 interfaces:
1. ITmdbEasyClient
2. IRequestHandler
3. IJsonSerializer

1. We absolutely need 1 one of these to be public, since we need a way to globally configure the library with options and a place to hold an HttpClient. We should therefore keep the TmdbEasyClient (class and interface) public and advise to register it as a singleton.

2. Hides this dependency. After some thought, i am not sure this class needs to be replaceable. We can allow for optional override via options

3. Hides this dependency, but allow an override via the options.

Changes made:
- Updated README with your suggested changes see [977ec7baa6a6572ec7abc080bc239490e77b415b]
- Renames `TmdbEasyClient` and removes version, see [d59a6445d83f5f8586d908220d011f607cec8a36]
- Moved urls into Constants class. I did this as I was taught to keep constant values in a Constants file. Seems neater and more organised IMHO, see [24840de9a50e9083321ce35f354a0c606cf98781] 
- Adds default `JsonSerializer`, allows for overriding with own implementation and marks internal, see [9e9fa134d5dc4fe1047ba867769d56524d20209c]
- Makes Request, Request

This means the following falls away:
``` csharp
// configuration
var deserializer = new NewtonSoftDeserializer();
var client = new TmdbEasyClientv3(deserializer, new TmdbEasyOptions());
var requestHandler = new RequestHandler(client);
// usage
var endPoint = new ConfigApi(requestHandler);
var jobs = await endPoint.GetJobsAsync();
``` 

and is replaced with this:
```csharp
// configuration
var client = new TmdbEasyClientv3(new TmdbEasyOptions());
// usage
var endPoint = new ConfigApi(client);
var jobs = await endPoint.GetJobsAsync();
```

**Comment by tonykaralis** [comment]
In all honesty I am not 200% sure if I want to merge the last change, hut will be the same

**Comment by PoLaKoSz** [comment]
Yes, sounds way better. Moving the deserializer to the options class was a great idea. :smile:

**Comment by tonykaralis** [comment]
FYI, I have released v1.0.0. Added tests for MovieApi and TelevisionApi. Listed a bug here #27. Have a good Sunday 👍

**Comment by PoLaKoSz** [comment]
Cool, have some rest now :smile:
